xbmc/interfaces/json-rpc/PVROperations.cpp

```diff
@@ -337,6 +337,103 @@ JSONRPC_STATUS CPVROperations::GetTimerDetails(const std::string &method, ITrans
   return OK;
 }
 
+JSONRPC_STATUS CPVROperations::AddTimer(const std::string &method, ITransportLayer *transport, IClient *client, const CVariant &parameterObject, CVariant &result)
+{
+  if (!g_PVRManager.IsStarted())
+    return FailedToExecute;
+
+  EpgSearchFilter filter;
+  filter.Reset();
+  filter.m_iUniqueBroadcastId = parameterObject["broadcastid"].asUnsignedInteger();
+
+  CFileItemList broadcasts;
+  int resultSize = g_EpgContainer.GetEPGSearch(broadcasts, filter);
+
+  if (resultSize != 1)
+    return InvalidRequest;
+
+  CFileItemPtr broadcast = broadcasts.Get(0);
+
+  if (!broadcast->HasEPGInfoTag())
+    return InvalidRequest;
+
+  CEpgInfoTagPtr epgTag = broadcast->GetEPGInfoTag();
+
+  bool repeating = parameterObject["repeating"].asBoolean(false);
+  if (epgTag->HasTimer())
+    return InvalidParams;
+
+  CPVRTimerInfoTagPtr newTimer = CPVRTimerInfoTag::CreateFromEpg(epgTag, repeating);
+  if (newTimer)
+  {
+    bool added = g_PVRTimers->AddTimer(newTimer);
+    if (added)
+    {
+      return ACK;
+    }
+  }
+  return FailedToExecute;
+}
+
+
+JSONRPC_STATUS CPVROperations::DeleteTimer(const std::string &method, ITransportLayer *transport, IClient *client, const CVariant &parameterObject, CVariant &result)
+{
+  if (!g_PVRManager.IsStarted())
+    return FailedToExecute;
+
+  CPVRTimers* timers = g_PVRTimers;
+
+  CPVRTimerInfoTagPtr timer = timers->GetById(parameterObject["timerid"].asInteger());
+  if (!timer)
+    return InvalidParams;
+
+  bool repeating = parameterObject["repeating"].asBoolean(false);
+
+  bool sentOkay = timers->DeleteTimer(timer, false, repeating);
+
+  if(sentOkay)
+    return ACK;
+
+  return FailedToExecute;
+}
+
+JSONRPC_STATUS CPVROperations::ToggleTimer(const std::string &method, ITransportLayer *transport, IClient *client, const CVariant &parameterObject, CVariant &result)
+{
+  if (!g_PVRManager.IsStarted())
+    return FailedToExecute;
+
+  EpgSearchFilter filter;
+  filter.Reset();
+  filter.m_iUniqueBroadcastId = parameterObject["broadcastid"].asUnsignedInteger();
+
+  CFileItemList broadcasts;
+  int resultSize = g_EpgContainer.GetEPGSearch(broadcasts, filter);
+
+  if (resultSize != 1)
+    return InvalidRequest;
+
+  CFileItemPtr broadcast = broadcasts.Get(0);
+
+  if (!broadcast->HasEPGInfoTag())
+    return InvalidRequest;
+
+  CEpgInfoTagPtr epgTag = broadcast->GetEPGInfoTag();
```

Comment:
The logic from the beginning of this method to this line is identical to the one in `AddTimer()` so maybe you could move it into a separate method which takes a `broadcastid` and returns `CEpgInfoTagPtr`.